bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java

```diff
@@ -0,0 +1,468 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.miio.internal.cloud;
+
+import java.io.IOException;
+import java.net.CookieStore;
+import java.net.HttpCookie;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.HttpResponseException;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.FormContentProvider;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.Fields;
+import org.openhab.binding.miio.internal.MiIoCrypto;
+import org.openhab.binding.miio.internal.MiIoCryptoException;
+import org.openhab.binding.miio.internal.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link MiCloudConnector} class is used for connecting to the Xiaomi cloud access
+ *
+ * @author Marcel Verpaalen - Initial contribution
+ */
+@NonNullByDefault
+public class MiCloudConnector {
+
+    private static final int REQUEST_TIMEOUT_SECONDS = 10;
+    private static final String UNEXPECTED = "Unexpected :";
+    private static final String AGENT_ID = (new Random().ints(65, 70).limit(13)
+            .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append).toString());
+    private static final String USERAGENT = "Android-7.1.1-1.0.0-ONEPLUS A3010-136-" + AGENT_ID
+            + " APP/xiaomi.smarthome APPV/62830";
+    private final JsonParser parser = new JsonParser();
+
+    // String
+    private String username;
+    private String password;
+
+    private final String clientId;
+    private String userId = "";
+    private String serviceToken = "";
```

Comment:
Why not use `Optional<String>` instead?

Reply:
I did not know about this Optional way, is indeed useful.
Though in this particular case I'm not clear what it will bring additionally as a empty string as servicetoken won't be helping me to logon. It would add test for isPresent() on top of the existing isempty ones